src/understand-tidb/the-lifecycle-of-a-statement.md

```diff
@@ -1 +1,200 @@
+
+## Executing the optimized plan
+
+The optimized plan is executed through [runStmt](https://github.com/pingcap/tidb/blob/30cf15a59db11c34ffe05fc926152a43327eaa61/session/session.go#L1750), which builds an [executor](https://github.com/pingcap/tidb/blob/30cf15a59db11c34ffe05fc926152a43327eaa61/executor/adapter.go#L319) from the plan and will return a record set or directly execute the statements in case no records will be returned, like `INSERT`/`UPDATE`/`DELETE` statements. Before returning the record set, the executor starts the execution by calling the Volcano inspired `Open()` API and the `Next()` api to retrieve the first chunk of data or execute the statement fully if no records are to be returned.
+
+The executors are often including coprocessors as seen above, where tasks can be seen as stream processors and can be parallelized and delegated to storage nodes (TiKV/TiFlash).
+
+### Requests sent to TiKV/TiFlash coprocessors
+
+During the execution different task are executed as coprocessors and delegated/pushed down to the storage nodes (TiKV/TiFlash) for both scaling and more optimized use of the cluster.
+
+This way there is less data sent between TiDB nodes and TiKV/TiFlash nodes (only filtered and aggregated results) and the computation/load are distributed across several storage nodes.
+
+Common coprocessors are: TableScan (simplest form no real optimisation), IndexScan (Range reads from index), Selection (Filter on condition, `WHERE` clause etc.), LIMIT (no more than N records), TopN (Order + Limit), Aggregation (`GROUP BY`)
+
+```go
+// HandleStreamRequest handles the coprocessor stream request.
+func (h *CoprocessorDAGHandler) HandleStreamRequest(ctx context.Context, req *coprocessor.Request, stream tikvpb.Tikv_CoprocessorStreamServer) error {
+  e, err := h.buildDAGExecutor(req)
+  err = e.Open(ctx)
+  chk := newFirstChunk(e)
+  for {
+    chk.Reset()
+    err = Next(ctx, e, chk)
+    // ...
+    if chk.NumRows() == 0 {
+       return h.buildResponseAndSendToStream(chk, ...)
+    }
+  }
+}
+```
+
+As seen above the Volcano inspired execution is iterating over `chunks` of data, not records one-by-one, which also allows for vectorization, which formats the chunk data so it can be processed as a vector instead of looping over each record and column one by one.
+
+### Sending the result back to the client
+
+If the statement returns a record set, it is handled in [writeChunks](https://github.com/pingcap/tidb/blob/30cf15a59db11c34ffe05fc926152a43327eaa61/server/conn.go#L1993) which loops over the record set's `Next()` until empty and then adds some context/metadata to the MySQL OK package and flushes the data back to the client.
+
+Notice that things like error handling, tracing etc. are not explained in this page.
```

Comment:
I think it would be useful to briefly allude to how DML changes will be buffered in a transaction buffer on the tidb-server. It can be covered in more detail under DML; but this is something that is different from MySQL.

Reply:
Thanks, I added it to the `Executing the optimized plan` section.